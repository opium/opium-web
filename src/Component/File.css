.MainContainer {
    display: flex;
    justify-content: center;
}

.ImageContainer {
    position: relative;
    display: flex;
    flex-direction: column;
    justify-content: center;
}

.Prev, .Next {
    position: absolute;
    top: 0;
    height: 100%;
    display: flex;
    flex-direction: column;
    justify-content: center;

    color: #fff;
    text-decoration: none;
}

.Prev:hover > .Overlay, .Next:hover > .Overlay {
    display: flex;
}

.Prev {
    left: 0;
    width: 30%;
    align-items: flex-start;
    padding-left: 30px;
}

.Next {
    right: 0;
    width: 45%;
    align-items: flex-end;
    padding-right: 30px;
}

.Overlay {
    background: rgba(0,0,0,.5);
    display: none;
    font-size: 60px;
    padding: 20px;
    border-radius: 5px;
}

.FileHeader {
    height: 80px;
    display: flex;
    align-items: center;
}

.FileTitle {
    flex: 2;
    text-align: center;
}

.BackContainer {
    flex-shrink: 0;
}

.BackToAlbum {
    margin-left: .5em;
    margin-right: .5em;
    padding: .8em;
    border-radius: 3px;
    font-size: 1.2em;
    background: #4F3A1B;
    color: #A0937F;
    text-decoration: none;
}

.BackToAlbum:hover {
    text-decoration: underline;
}

.Image {
    display: inline;
    max-width: 100%;
    max-height: 100%;
}

.Table {
    margin: 20px 0;
    width: 100%;
    max-width: 100%;
    border-spacing: 0;
    border-collapse: collapse;
}

.Table th, .Table td {
    padding: 8px;
    line-height: 1.42857143;
    vertical-align: top;
    border-top: 1px solid #ddd;
}
.Table th {
    text-align: left;
}
