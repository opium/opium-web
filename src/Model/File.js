import { Map, Record } from 'immutable';
import { mapEntityRelationShips } from './Factory';

class File extends Record({
  id: null,
  name: null,
  slug: null,
  thumbnails: null,
}) {
  constructor(val) {
    const data = val;
    data.thumbnails = Map(val.thumbnails);

    return mapEntityRelationShips(super(data), data);
  }
}

export default File;
