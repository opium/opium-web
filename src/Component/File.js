import React, { Component, PropTypes } from 'react';
import { Link } from 'react-router';
import Helmet from 'react-helmet';
import ChevronLeft from 'react-icons/lib/ti/chevron-left';
import ChevronRight from 'react-icons/lib/ti/chevron-right';
import FileModel from '../Model/File';
import { loadImage } from '../ImageLoader';
import Loader from './Loader';
import './File.css';

const BackLink = ({file, ...props}) => {
  if (!file.parent) {
    return null;
  }

  return (
    <Link to={`/${file.parent.slug}`} {...props}>
      <ChevronLeft /> {file.parent.name}
    </Link>
  );
};

const PrevLink = ({file}) => {
  if (!file.previous) {
    return null;
  }

  return (
    <Link className="Prev" to={`/${file.parent.slug}/${file.previous.slug}`}>
      <div className="Overlay">
        <ChevronLeft />
      </div>
    </Link>
  );
};

const NextLink = ({file}) => {
  if (!file.next) {
    return null;
  }

  return (
    <Link className="Next" to={`/${file.parent.slug}/${file.next.slug}`}>
      <div className="Overlay">
        <ChevronRight />
      </div>
    </Link>
  );
};

const Exif = ({file}) =>
  <table className="Table">
    <tbody>
      <tr>
        <th>Camera</th>
        <td>
          {file.exif.get('Make')}
          {' '}
          {file.exif.get('Model')}
        </td>
      </tr>
      <tr>
        <th>Date</th>
        <td>
          {file.exif.get('DateTime')}
        </td>
      </tr>
      <tr>
        <th>Size</th>
        <td>
          {file.exif.get('ExifImageWidth')}
          x
          {file.exif.get('ExifImageLength')}
        </td>
      </tr>
    </tbody>
  </table>
;

class File extends Component {
  static propTypes = {
    file: PropTypes.instanceOf(FileModel).isRequired,
    isFetchingFile: PropTypes.bool.isRequired,
    findFile: PropTypes.func.isRequired,
    removeCurrentFile: PropTypes.func.isRequired,
    viewportHeight: PropTypes.number.isRequired,
    pushLocation: PropTypes.func.isRequired,
  }

  constructor(props) {
    super(props);

    this.handleKeyup = this.handleKeyup.bind(this);

    this.state = {
      imageLoaded: false,
    };
  }

  componentDidMount() {
    this.props.findFile(this.props.slug);

    window.addEventListener('keyup', this.handleKeyup);
  }

  componentDidUpdate(prevProps) {
    if (this.props.slug !== prevProps.slug) {
      this.setState({ imageLoaded: false });
      this.props.findFile(this.props.slug);
    }

    if (this.props.file !== prevProps.file) {
      this.loadFileImage();
    }
  }

  handleKeyup(e) {
    const file = this.props.file;

    switch (e.keyCode) {
      case 37:
        if (file.previous) {
          this.props.pushLocation(`/${file.parent.slug}/${file.previous.slug}`);
        }
        break;
      case 39:
        if (file.next) {
          this.props.pushLocation(`/${file.parent.slug}/${file.next.slug}`);
        }
        break;
      default:
        break;
    }
  }

  loadFileImage() {
    const file = this.props.file;
    const src = file.thumbnails.get('image');

    if (!src) {
      return;
    }

    loadImage(src).then(() => this.setState({ imageLoaded: true }));
  }

  componentWillUnmount() {
    this.props.removeCurrentFile();
  }

  render() {
    const file = this.props.file;

    if (!file.thumbnails.get('image') || this.props.isFetchingFile) {
      return <div />;
    }

    return (
      <div>
        <Helmet title={file.name} />

        <header className="FileHeader">
          <div>
            <BackLink file={file} className="BackToAlbum" />
          </div>
          <h1 className="FileTitle">
            {file.name}
          </h1>
        </header>

        <div className="MainContainer">
          <div
            className="ImageContainer"
            style={{ height: `${this.props.viewportHeight - 80}px` }}
          >
            <PrevLink file={file} />
            <NextLink file={file} />

            {this.state.imageLoaded ?
                <img
                  src={file.thumbnails.get('image')}
                  alt={file.name}
                  className="Image"
                /> :
              <Loader color="#594F3F" />
            }
          </div>
        </div>

        <Exif file={file} />
      </div>
    );
  }
}

export default File;
