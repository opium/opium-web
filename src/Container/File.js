import { connect } from 'react-redux';
import { push } from 'react-router-redux';
import FileComponent from '../Component/File';
import File from '../Model/File';
import { find, removeCurrent } from '../Action/FileAction';

function mapStateToProps(state, ownProps) {
  return {
    file: state.opium.get('currentFile') || new File({}),
    isFetchingFile: !!state.opium.get('isFetchingFile'),
    slug: ownProps.params.fileSlug || '',
    viewportHeight: document.documentElement.clientHeight,
  }
};

const mapDispatchToProps = {
  findFile: find,
  removeCurrentFile: removeCurrent,
  pushLocation: push,
};

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(FileComponent);
