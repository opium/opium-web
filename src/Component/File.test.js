import React from 'react';
import renderer from 'react-test-renderer';
import File from './File';
import FileModel from '../Model/File';

jest.mock('react-leaflet');

it('renders empty file without crashing', () => {
  const component = renderer.create(<File
    file={new FileModel({})}
    findFile={() => {}}
    removeCurrentFile={() => {}}
    pushLocation={() => {}}
    isFetchingFile={false}
    viewportHeight={700}
  />);
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot()
});

it('renders a loader if file is not loaded', () => {
  const file = new FileModel({});
  const component = renderer.create(<File
    file={file}
    findFile={() => {}}
    removeCurrentFile={() => {}}
    pushLocation={() => {}}
    isFetchingFile={true}
    viewportHeight={700}
  />);
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot()
});

it('renders real file without crashing', () => {
  const file = new FileModel({
    name: 'Foo',
    slug: 'foo',
    thumbnails: {
      image: 'http://demo.opium.sitioweb.fr/2010-mavericks-competition-edit1-jpg',
    },
  });
  const component = renderer.create(<File
    file={file}
    findFile={() => {}}
    removeCurrentFile={() => {}}
    pushLocation={() => {}}
    isFetchingFile={false}
    viewportHeight={700}
  />);
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot()
});

it('renders file with backlink', () => {
  const file = new FileModel({
    name: 'Foo',
    slug: 'foo',
    thumbnails: {
      image: 'http://demo.opium.sitioweb.fr/2010-mavericks-competition-edit1-jpg',
    },
    parent: {
      name: 'Bar',
      slug: 'bar',
    },
  });
  const component = renderer.create(<File
    file={file}
    findFile={() => {}}
    removeCurrentFile={() => {}}
    pushLocation={() => {}}
    isFetchingFile={false}
    viewportHeight={700}
  />);
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot()
});

it('renders file with prev / next links', () => {
  const file = new FileModel({
    name: 'Foo',
    slug: 'foo',
    thumbnails: {
      image: 'http://demo.opium.sitioweb.fr/2010-mavericks-competition-edit1-jpg',
    },
    parent: {
      name: 'Directory',
      slug: 'directory',
    },
    previous: {
      name: 'Bar',
      slug: 'bar',
    },
    next: {
      name: 'Baz',
      slug: 'baz',
    },
  });
  const component = renderer.create(<File
    file={file}
    findFile={() => {}}
    removeCurrentFile={() => {}}
    pushLocation={() => {}}
    isFetchingFile={false}
    viewportHeight={700}
  />);
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot()
});

it('renders file with a map', () => {
  const file = new FileModel({
    name: 'Foo',
    slug: 'foo',
    thumbnails: {
      image: 'http://demo.opium.sitioweb.fr/2010-mavericks-competition-edit1-jpg',
    },
    parent: {
      name: 'Directory',
      slug: 'directory',
    },
    position: {
      lat: 45,
      lng: 4,
    },
  });
  const component = renderer.create(<File
    file={file}
    findFile={() => {}}
    removeCurrentFile={() => {}}
    pushLocation={() => {}}
    isFetchingFile={false}
    viewportHeight={700}
  />);
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot()
});
