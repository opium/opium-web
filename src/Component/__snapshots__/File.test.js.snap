exports[`test renders a loader if file is not loaded 1`] = `<div />`;

exports[`test renders empty file without crashing 1`] = `<div />`;

exports[`test renders file with backlink 1`] = `
<div>
  <header
    className="FileHeader">
    <div>
      <a
        className="BackToAlbum"
        onClick={[Function]}
        style={Object {}}>
        <svg
          fill="currentColor"
          height="1em"
          preserveAspectRatio="xMidYMid meet"
          style={
            Object {
              "color": undefined,
              "verticalAlign": "middle",
            }
          }
          viewBox="0 0 40 40"
          width="1em">
          <g>
            <path
              d="m24 9.3c-1.3-1.3-3.4-1.3-4.7 0l-10.7 10.7 10.7 10.7c0.7 0.6 1.5 1 2.4 1s1.7-0.4 2.3-1c1.3-1.3 1.3-3.4 0-4.7l-5.9-6 5.9-6c1.3-1.3 1.3-3.4 0-4.7z" />
          </g>
        </svg>
         
        Bar
      </a>
    </div>
    <h1
      className="FileTitle">
      Foo
    </h1>
  </header>
  <div
    className="MainContainer">
    <div
      className="ImageContainer"
      style={
        Object {
          "height": "620px",
        }
      }>
      <div
        className="loader">
        <div
          className="ball-scale-ripple-multiple">
          <div
            style={
              Object {
                "borderColor": "#594F3F",
              }
            } />
          <div
            style={
              Object {
                "borderColor": "#594F3F",
              }
            } />
          <div
            style={
              Object {
                "borderColor": "#594F3F",
              }
            } />
        </div>
      </div>
    </div>
  </div>
  <table
    className="Table">
    <tbody>
      <tr>
        <th>
          Camera
        </th>
        <td>
           
        </td>
      </tr>
      <tr>
        <th>
          Date
        </th>
        <td />
      </tr>
      <tr>
        <th>
          Size
        </th>
        <td>
          x
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;

exports[`test renders file with prev / next links 1`] = `
<div>
  <header
    className="FileHeader">
    <div>
      <a
        className="BackToAlbum"
        onClick={[Function]}
        style={Object {}}>
        <svg
          fill="currentColor"
          height="1em"
          preserveAspectRatio="xMidYMid meet"
          style={
            Object {
              "color": undefined,
              "verticalAlign": "middle",
            }
          }
          viewBox="0 0 40 40"
          width="1em">
          <g>
            <path
              d="m24 9.3c-1.3-1.3-3.4-1.3-4.7 0l-10.7 10.7 10.7 10.7c0.7 0.6 1.5 1 2.4 1s1.7-0.4 2.3-1c1.3-1.3 1.3-3.4 0-4.7l-5.9-6 5.9-6c1.3-1.3 1.3-3.4 0-4.7z" />
          </g>
        </svg>
         
        Directory
      </a>
    </div>
    <h1
      className="FileTitle">
      Foo
    </h1>
  </header>
  <div
    className="MainContainer">
    <div
      className="ImageContainer"
      style={
        Object {
          "height": "620px",
        }
      }>
      <a
        className="Prev"
        onClick={[Function]}
        style={Object {}}>
        <div
          className="Overlay">
          <svg
            fill="currentColor"
            height="1em"
            preserveAspectRatio="xMidYMid meet"
            style={
              Object {
                "color": undefined,
                "verticalAlign": "middle",
              }
            }
            viewBox="0 0 40 40"
            width="1em">
            <g>
              <path
                d="m24 9.3c-1.3-1.3-3.4-1.3-4.7 0l-10.7 10.7 10.7 10.7c0.7 0.6 1.5 1 2.4 1s1.7-0.4 2.3-1c1.3-1.3 1.3-3.4 0-4.7l-5.9-6 5.9-6c1.3-1.3 1.3-3.4 0-4.7z" />
            </g>
          </svg>
        </div>
      </a>
      <a
        className="Next"
        onClick={[Function]}
        style={Object {}}>
        <div
          className="Overlay">
          <svg
            fill="currentColor"
            height="1em"
            preserveAspectRatio="xMidYMid meet"
            style={
              Object {
                "color": undefined,
                "verticalAlign": "middle",
              }
            }
            viewBox="0 0 40 40"
            width="1em">
            <g>
              <path
                d="m14.3 9.3c-1.3 1.3-1.3 3.4 0 4.7l6 6-6 6c-1.3 1.3-1.3 3.4 0 4.7 0.7 0.6 1.5 1 2.4 1s1.7-0.4 2.3-1l10.7-10.7-10.7-10.7c-1.3-1.3-3.4-1.3-4.7 0z" />
            </g>
          </svg>
        </div>
      </a>
      <div
        className="loader">
        <div
          className="ball-scale-ripple-multiple">
          <div
            style={
              Object {
                "borderColor": "#594F3F",
              }
            } />
          <div
            style={
              Object {
                "borderColor": "#594F3F",
              }
            } />
          <div
            style={
              Object {
                "borderColor": "#594F3F",
              }
            } />
        </div>
      </div>
    </div>
  </div>
  <table
    className="Table">
    <tbody>
      <tr>
        <th>
          Camera
        </th>
        <td>
           
        </td>
      </tr>
      <tr>
        <th>
          Date
        </th>
        <td />
      </tr>
      <tr>
        <th>
          Size
        </th>
        <td>
          x
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;

exports[`test renders real file without crashing 1`] = `
<div>
  <header
    className="FileHeader">
    <div />
    <h1
      className="FileTitle">
      Foo
    </h1>
  </header>
  <div
    className="MainContainer">
    <div
      className="ImageContainer"
      style={
        Object {
          "height": "620px",
        }
      }>
      <div
        className="loader">
        <div
          className="ball-scale-ripple-multiple">
          <div
            style={
              Object {
                "borderColor": "#594F3F",
              }
            } />
          <div
            style={
              Object {
                "borderColor": "#594F3F",
              }
            } />
          <div
            style={
              Object {
                "borderColor": "#594F3F",
              }
            } />
        </div>
      </div>
    </div>
  </div>
  <table
    className="Table">
    <tbody>
      <tr>
        <th>
          Camera
        </th>
        <td>
           
        </td>
      </tr>
      <tr>
        <th>
          Date
        </th>
        <td />
      </tr>
      <tr>
        <th>
          Size
        </th>
        <td>
          x
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;
